### PR TITLE
LPS-154537 Implement our own `memoize` in order to replace `lodash.memoize`

### PR DESCRIPTION
**Pull request by kresimir-coko**
I haven't done extensive testing of this util, so I'm making this a draft PR to see what people have to say about the implementation first before I commit time to testing it in DXP.

### CODE REVIEW

**Comment by kresimir-coko** [comment]
ci:reevaluate:1397401_543

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		const memoizedMockFn = memoize(mockFn);
+
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+
+		expect(memoizedMockFn.getCache().size).toBe(1);
+	});
+
+	it("throws an error if the provided argument isn't a function", () => {
+		expect(() => memoize({foo: 42})).toThrow(TypeError);
+	});
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:53]: Can we add some more functional type tests here to test the function of memoize? For example, 

- Ensuring that the provided function is only called once, even if the memo is invoked twice
- Ensuring the value is equal between two calls of the memo'd function
- Ensure that the same object is returned if a memo is called twice with same args

For example
```js
function retObj(a,b) {
    return {a,b}
}

retObj('foo', 'bar') !== retObj('foo', 'bar') // it would create new object each time

memoizedRetObj('foo', 'bar') === memoizedRetObj('foo', 'bar') // should return the same object
```

I'd also grab some of the tests from the lodash unit tests and add them here to make sure we aren't missing any edge cases

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			number,
+			array,
+			string,
+		]);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+
+		expect(memoizedMockFn.getCache().size).toBe(1);
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:48]: Checking the internals of the memo isn't super helpful here. It is best to test the expected return of the memo

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js (hunk 1 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:53]: Great suggestions, will update next week, thank you @bryceosterhaus

**Comment by kresimir-coko** [comment]
Hey @bryceosterhaus I've moved the variable declarations used across multiple tests outside of the test descriptions to keep the code DRY. 

> Ensuring the value is equal between two calls of the memo'd function
> Ensure that the same object is returned if a memo is called twice with same args

I combined your 2 points into one test testing the object return -> `returns a cached value if a function is invoked with same inputs`

As for the lodash tests you referred to, I [checked them out](https://github.com/lodash/lodash/blob/master/test/memoize.test.js) but they are mostly overlapping with ours or testing some specific cache functionality that they are utilizing.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		const mockFn = jest.fn((object) => object);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		memoizedMockFn(mockObject);
+		memoizedMockFn(mockObjectTwo);
+		memoizedMockFn(mockObject);
+
+		expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));
+		expect(memoizedMockFn(mockObject)).not.toEqual(
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:65]: I think this would be better written as

```js
// This ensures the memo works
expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));

// This ensures that even though the keys and values in the object are the same, the object itself is not the same reference.
expect(memoizedMockFn(mockObject)).not.toEqual(mockFn(mockObject));
```

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+			12: 24,
+			array: [1, 2, '3'],
+			foo: 'bar',
+		};
+
+		const mockObjectTwo = {
+			array: [4, '5', 6],
+			bar: 'foo',
+			memo: 'ize',
+		};
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:54]: I don't think this obj is necessary with my suggestions below

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+			['foo', 'bar'],
+			'baz',
+		]);
+	});
+
+	it('calls the provided function once, even if the memo is invoked twice', () => {
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+		memoizedMockFn({3: 4}, 4, ['foo', 'bar'], 'baz');
+
+		expect(mockFn).toHaveBeenCalledTimes(1);
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:40]: see above comment

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+import memoize from '../../../src/main/resources/META-INF/resources/liferay/util/memoize';
+
+const mockFn = jest.fn((object, number, array, string) => [
+	object,
+	number,
+	array,
+	string,
+]);
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:22]: Probably simpler to write. I think this would work..

```suggestion
const mockFn = jest.fn((...args) => [...args]);
```

**Review by bryceosterhaus — CHANGES_REQUESTED** [review]

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			memo: 'ize',
+		};
+
+		const mockFn = jest.fn((object) => object);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		memoizedMockFn(mockObject);
+		memoizedMockFn(mockObjectTwo);
+		memoizedMockFn(mockObject);
```
[the remark this answers is not in the document]
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:62]: Between all the comments and ideas it was left as a test of caching behaviour but with your suggestions, it ended up unnecessary.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js (hunk 3 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:65]: @bryceosterhaus I think [toBe](https://jestjs.io/docs/expect#tobevalue) is the better avenue for testing object reference/value here, so I'll go with that.

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:65]: ❌ `expect(memoizedMockFn(mockObject)).not.toEqual(mockFn(mockObject));` doesn't pass

@bryceosterhaus  It seems that using the `mockObject` variable maintains the reference to the same object, should I use the approach below?

✅  below comparison does pass
```js
expect(
	memoizedMockFn({
		12: 24,
		array: [1, 2, '3'],
		foo: 'bar',
	})
).not.toBe(
	mockFn({
		12: 24,
		array: [1, 2, '3'],
		foo: 'bar',
	})
);
```

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:65]: I also can't figure out what is the reason for this test, like, what should I put in the `it('?', () => ...)`

I wanted this test to test that it's using a cached value. So I can't find a connection between your suggestion and that goal.

Comment (bryceosterhaus) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js]: >  It seems that using the mockObject variable maintains the reference to the same object

Oh yeah that is true, so use the example you provided.

> I wanted this test to test that it's using a cached value.

I think this would still be accurate because you are testing that the value is returning the cached value in this tests
```js
expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));
```

But then you are also testing the inverse and making sure that the memo'd version doesn't return the same as the non-memo version. This could technically be written as it's own unit test, but I think it works here too

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js (hunk 5 above)
Comment (bryceosterhaus) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:39]: Your most recent changes fix this though, so no need to worry

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+			array: [1, 2, '3'],
+			foo: 'bar',
+		};
+
+		const mockFn = jest.fn((object) => object);
+
+		const memoizedMockFn = memoize(mockFn);
+
+		expect(memoizedMockFn(mockObject)).toEqual(memoizedMockFn(mockObject));
+	});
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:54]: Thinking about your comment about the reference to `mockObject`, I think this test would also pass without using the memo, because it would reference `mockObject` in both invocations.

I would write this test like

```js
	it('returns the same value if invoked with same arguments', () => {
		const mockFn = jest.fn((val) => val);

		const memoizedMockFn = memoize(mockFn);

		expect(memoizedMockFn({foo: 'bar'})).toEqual(memoizedMockFn({foo: 'bar'}));
	});
```

**Review by bryceosterhaus — CHANGES_REQUESTED** [review]

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js (hunk 8 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/util/memoize.js:54]: Yeah, that's where I ended up at but was waiting for your responses before pushing 😉

**Comment by kresimir-coko** [comment]
@bryceosterhaus I pushed the update

**Comment by kresimir-coko** [comment]
ci:test:sf

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
Reviewed. Test failures are unrelated. :heavy_check_mark: 

Ok to manually forward from QA perspective. Pending dev approval.

[Testray Comparison](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=3&testrayRunIdB=2004673109&testrayRunIdA=2004657229&statusB=2)

The test failures seem to be related to a Commerce/Core race condition affecting upstream and looks like being worked on in [this thread](https://liferay.slack.com/archives/CLCD3DQLF/p1665519831787109)

**Comment by bryceosterhaus** [comment]
manual forward https://github.com/brianchandotcom/liferay-portal/pull/124499
